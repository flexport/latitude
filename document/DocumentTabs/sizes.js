/**
 * TEAM: frontend_infra
 * @flow strict
 */

/* eslint-disable import/prefer-default-export */

const sizes = Object.freeze({
  s: 40,
  m: 56,
  l: 56,
});

export default sizes;
export type Size = $Keys<typeof sizes>;
