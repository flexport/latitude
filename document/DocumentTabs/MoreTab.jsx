/**
 * TEAM: frontend_infra
 * @flow
 */

import * as React from "react";
import {StyleSheet, css} from "aphrodite";
import Dotdotdot from "react-dotdotdot";
import PopupWithClickAway from "../../popup/PopupWithClickAway";
import Text from "../../Text";
import Icon from "../../Icon";
import colors from "../../colors";
import Clickable from "../../base_candidate/Clickable";
import typeof DocumentTab from "./DocumentTab";
import {type Size} from "./sizes";

const HORIZONTAL_MARGIN = 4;

type Props = {|
  +children: React.ChildrenArray<React.Element<DocumentTab>>,
  +width: number,
  +selectedLabel: string | null,
  +size: Size,
|};

export const ListPopupContext = React.createContext<{
  togglePopup: () => void,
}>({
  togglePopup: () => {},
});

function MoreTab({width, selectedLabel, size, children}: Props) {
  const marginAdjustedWidth = width - HORIZONTAL_MARGIN * 2;

  return (
    <div className={css(styles.container)} style={{width: marginAdjustedWidth}}>
      <PopupWithClickAway>
        {(Target, Popup, {togglePopup, isOpen}) => (
          <>
            <Target>
              <Clickable onClick={togglePopup}>
                <div
                  className={css(
                    styles.button,
                    !!selectedLabel && styles.selected
                  )}
                  style={selectedLabel && {width: marginAdjustedWidth}}
                >
                  {selectedLabel ? (
                    <div className={css(styles.labelWrapper)}>
                      <Dotdotdot clamp={size !== "s" ? 2 : 1}>
                        {selectedLabel}
                      </Dotdotdot>
                    </div>
                  ) : (
                    <div className={css(styles.textWrapper)}>
                      <Text color="white" weight="bold">
                        More
                      </Text>
                    </div>
                  )}
                  <Icon iconName="downOpen" color="white" size="xs" />
                </div>
              </Clickable>
            </Target>
            {isOpen ? (
              <Popup placement="bottom-end">
                <div
                  className={css(styles.listContainer)}
                  style={{width: marginAdjustedWidth}}
                >
                  <ListPopupContext.Provider value={{togglePopup}}>
                    {children}
                  </ListPopupContext.Provider>
                </div>
              </Popup>
            ) : null}
          </>
        )}
      </PopupWithClickAway>
    </div>
  );
}

const styles = StyleSheet.create({
  container: {
    minHeight: "100%",
    display: "flex",
    position: "relative",
    margin: `0 ${HORIZONTAL_MARGIN}px`,
    justifyContent: "flex-end",
    alignItems: "stretch",
  },
  button: {
    display: "flex",
    padding: "6px 8px",
    minHeight: "100%",
    alignItems: "center",
    ":hover": {
      backgroundColor: colors.grey60,
    },
  },
  textWrapper: {
    margin: "0 8px 0 4px",
    boxShadow: "inset 0 -2px 0 0 #FFF",
  },
  labelWrapper: {
    wordWrap: "break-word",
    fontSize: "12px",
    lineHeight: "16px",
    color: colors.white,
    flex: "0 1 auto",
    overflow: "hidden",
    position: "relative",
    marginRight: "4px",
  },
  listContainer: {
    marginTop: "4px",
    minWidth: "100%",
  },
  selected: {
    backgroundColor: colors.grey50,
  },
});

export default MoreTab;
