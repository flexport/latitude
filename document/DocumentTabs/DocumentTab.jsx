/**
 * TEAM: frontend_infra
 * @flow
 */

import * as React from "react";
import {StyleSheet, css} from "aphrodite";
import Dotdotdot from "react-dotdotdot";
import PopupWithClickAway from "../../popup/PopupWithClickAway";
import GeneralPopover from "../../popover/GeneralPopover";
import Button from "../../button/Button";
import IconButton from "../../button/IconButton";
import Tooltip from "../../Tooltip";
import Text from "../../Text";
import colors from "../../colors";
import {type Size} from "./sizes";
import {ListPopupContext} from "./MoreTab";
import Clickable from "../../base_candidate/Clickable";

const HORIZONTAL_MARGIN = 4;

type Props = {|
  +children: string,
  +isOverflow: boolean,
  +width: number,
  +isSelected: boolean,
  +isInDropdown: boolean,
  +onClick: () => void,
  +onDelete: () => void,
  +size: Size,
|};

function DocumentTab({
  children,
  isOverflow,
  width,
  size,
  isSelected,
  onClick,
  onDelete,
  isInDropdown,
}: Props) {
  const listPopupContext = React.useContext(ListPopupContext);
  const marginAdjustedWidth = width - HORIZONTAL_MARGIN * 2;

  return (
    <PopupWithClickAway>
      {(Target, Popup, {togglePopup, isOpen}) => (
        <>
          <Target>
            <Tooltip
              placement={isOverflow ? "right" : "bottom"}
              mouseEnterDelay={0.5}
              overlay={
                <div
                  style={{
                    minWidth: marginAdjustedWidth,
                    wordWrap: "break-word",
                  }}
                >
                  {children}
                </div>
              }
            >
              <Clickable
                onClick={() => {
                  onClick();
                  if (!isOverflow) {
                    listPopupContext.togglePopup();
                  }
                }}
              >
                <div
                  className={css(
                    styles.container,
                    (isSelected || isOpen) && styles.selected,
                    isInDropdown && styles.dropdownContainer,
                    isOverflow && isSelected && styles.dropdownSelected
                  )}
                  style={{width: marginAdjustedWidth}}
                >
                  <div className={css(styles.textContainer)}>
                    <Dotdotdot clamp={size !== "s" ? 2 : 1}>
                      {children}
                    </Dotdotdot>
                  </div>

                  <div
                    className={css(
                      styles.buttonContainer,
                      !isSelected && !isOpen && styles.hidden
                    )}
                  >
                    <IconButton
                      iconName="trash"
                      type="button"
                      intent="none"
                      kind="blank"
                      onClick={e => {
                        e.stopPropagation();
                        togglePopup();
                      }}
                    />
                  </div>
                </div>
              </Clickable>
            </Tooltip>
          </Target>
          <Popup placement={isOverflow ? "right-start" : "bottom-start"}>
            <div style={{margin: isOverflow ? "0 8px" : "8px 0"}}>
              <GeneralPopover
                title="Delete Shipping Order"
                buttons={[
                  <Button key={1} label="Cancel" onClick={togglePopup} />,
                  <Button
                    key={2}
                    label="Delete"
                    intent="danger"
                    kind="solid"
                    onClick={() => {
                      onDelete();
                      togglePopup();
                    }}
                  />,
                ]}
              >
                <div className={css(styles.popoverBody)}>
                  <Text display="inline">
                    Are you sure you want to delete <b>{children}</b>?
                  </Text>
                  <br />
                  <Text display="inline">
                    You will not be able to access this file once it is deleted.
                    Data from this file will be deleted as well.
                  </Text>
                  <br />
                </div>
              </GeneralPopover>
            </div>
          </Popup>
        </>
      )}
    </PopupWithClickAway>
  );
}

const styles = StyleSheet.create({
  container: {
    display: "flex",
    position: "relative",
    alignItems: "center",
    padding: "6px 8px",
    margin: `0 ${HORIZONTAL_MARGIN}px`,
    justifyContent: "space-between",
    cursor: "pointer",
    minHeight: "100%",
    backgroundColor: colors.black,
    color: colors.white,
    ":hover": {
      backgroundColor: colors.grey60,
    },
    ":hover > :nth-child(2)": {
      visibility: "visible",
    },
    ":active": {
      backgroundColor: colors.grey40,
    },
  },
  dropdownContainer: {
    padding: "8px 12px",
    backgroundColor: colors.grey60,
    margin: "0",
    ":hover": {
      backgroundColor: colors.grey50,
    },
  },
  selected: {
    backgroundColor: colors.grey50,
  },
  textContainer: {
    wordWrap: "break-word",
    fontSize: "12px",
    lineHeight: "16px",
    overflow: "hidden",
    flex: "0 1 auto",
  },
  buttonContainer: {
    justifyContent: "center",
    height: "100%",
    margin: "0 -4px 0 4px",
    padding: "4px",
    // Hack: since inverted colors for buttons are not yet supported
    ":nth-child(1n) > button": {
      fill: colors.grey20,
      ":hover span svg": {
        fill: colors.white,
      },
    },
  },
  hidden: {
    visibility: "hidden",
  },
  dropdownSelected: {
    background: colors.black,
  },
  popoverBody: {
    display: "flex",
    flexDirection: "column",
    wordWrap: "break-word",
  },
});

export default DocumentTab;
