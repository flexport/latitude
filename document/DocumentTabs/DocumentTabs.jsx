/**
 * TEAM: frontend_infra
 * @flow
 */

import * as React from "react";
import {StyleSheet, css} from "aphrodite";
import colors from "../../colors";
import DocumentTab from "./DocumentTab";
import MoreTab from "./MoreTab";
import IconButton from "../../button/IconButton";
import type {Size} from "./sizes";

const MIN_ELEMENT_WIDTH = 130;
const MAX_ELEMENT_WIDTH = 200;

type Props = {|
  /** A list of document names to be displayed. */
  +documentNameOptions: $ReadOnlyArray<{+key: number, +name: string}>,
  /** The key of the currently selected document. */
  +selectedKey: number,
  /** A callback called when the add button is pressed. */
  +onAdd: () => void,
  /** A callback called when a tab is selected. */
  +onSelect: (key: number) => void,
  /** A callback called when a document is deleted. */
  +onDelete: (key: number) => void,
  /** The size of document tabs. */
  +size?: Size,
|};

type Dimensions = {|
  tabWidth: number,
  slices: number,
|};

function calculateWidths(
  containerWidth: number,
  minWidth: number,
  maxWidth: number,
  elementCount: number
): Dimensions {
  if (elementCount * maxWidth < containerWidth) {
    return {tabWidth: maxWidth, slices: elementCount};
  }

  const maxSlicesCount = Math.floor(containerWidth / minWidth);
  const slices = Math.min(maxSlicesCount, elementCount);
  // slices will be 0 sometimes.
  const tabWidth = slices !== 0 ? containerWidth / slices : 0;

  return {
    tabWidth,
    slices,
  };
}

/**
 * @short A tab header for managing multiple documents
 * @brandStatus V2
 * @status Beta
 * @category Documents
 */
function DocumentTabs({
  documentNameOptions,
  selectedKey,
  onAdd,
  onSelect,
  onDelete,
  size = "m",
}: Props) {
  const containerRef = React.useRef<HTMLDivElement | null>(null);
  const [dimensions, setDimensions] = React.useState({
    tabWidth: 0,
    slices: 0,
  });
  const {tabWidth, slices} = dimensions;

  const offsetWidth = containerRef?.current?.offsetWidth || 0;

  React.useEffect(() => {
    if (containerRef.current !== null) {
      const dimensions = calculateWidths(
        // container size offset by size of addition button
        containerRef.current.offsetWidth,
        MIN_ELEMENT_WIDTH,
        MAX_ELEMENT_WIDTH,
        documentNameOptions.length
      );

      setDimensions(dimensions);
    }
  }, [documentNameOptions, offsetWidth]);

  const isOverflowing = documentNameOptions.length > slices;
  const nonOverflowCount = isOverflowing ? slices - 1 : slices;

  return (
    <div
      className={css(styles.container, tabWidth === 0 && styles.hidden)}
      style={{display: documentNameOptions.length !== 0 ? "flex" : "none"}}
    >
      <div className={css(styles.tabsContainer)} ref={containerRef}>
        {documentNameOptions.slice(0, nonOverflowCount).map(({name, key}) => (
          <DocumentTab
            isOverflow={false}
            width={tabWidth}
            isSelected={key === selectedKey}
            onClick={() => onSelect(key)}
            onDelete={() => onDelete(key)}
            size={size}
            isInDropdown={false}
            key={key}
          >
            {name}
          </DocumentTab>
        ))}

        {isOverflowing ? (
          <MoreTab
            width={tabWidth}
            size={size}
            selectedLabel={
              selectedKey < nonOverflowCount
                ? null
                : (
                    documentNameOptions.find(({key}) => key === selectedKey) ||
                    {}
                  ).name
            }
          >
            {documentNameOptions.slice(nonOverflowCount).map(({name, key}) => (
              <DocumentTab
                isOverflow={true}
                width={tabWidth}
                isSelected={key === selectedKey}
                onClick={() => onSelect(key)}
                onDelete={() => onDelete(key)}
                isInDropdown={true}
                size={size}
                key={key}
              >
                {name}
              </DocumentTab>
            ))}
          </MoreTab>
        ) : (
          undefined
        )}
      </div>

      <div className={css(styles.addButtonWrapper)}>
        <IconButton
          iconName="upload"
          type="button"
          intent="none"
          kind="bare"
          onClick={onAdd}
        />
      </div>
    </div>
  );
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: colors.black,
    padding: "12px 28px",
    flexDirection: "row",
    width: "100%",
    alignItems: "stretch",
  },
  tabsContainer: {
    flexDirection: "row",
    display: "flex",
    flex: "1 1 auto",
    alignItems: "stretch",
    padding: "2px 0",
    overflow: "hidden",
  },
  addButtonWrapper: {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    margin: "0 4px 0 28px",
    // Hack: since inverted colors for buttons are not yet supported
    ":nth-child(1n) > button": {
      fill: colors.white,
      ":hover": {
        background: colors.grey60,
      },
      ":hover span svg": {
        fill: colors.white,
      },
    },
  },
  hidden: {
    visibility: "hidden",
  },
});

export const test = {calculateWidths};
export default DocumentTabs;
