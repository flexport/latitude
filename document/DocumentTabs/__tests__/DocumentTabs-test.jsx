/**
 * TEAM: frontend_infra
 *
 * @flow
 */

import React from "react";
import {shallow} from "enzyme";
import DocumentTabs, {test} from "../DocumentTabs";

describe("DocumentTabs", () => {
  it("renders", () => {
    expect(
      shallow(
        <DocumentTabs
          documentNameOptions={["doc1.pdf", "doc2.pdf"].map((name, key) => ({
            name,
            key,
          }))}
          selectedKey={0}
          onSelect={() => {}}
          onAdd={() => {}}
          onDelete={() => {}}
        />
      )
    ).toMatchSnapshot();
  });
  it("calculateWidths will handle zero slices scenario", () => {
    const containerWidth = 99;
    const elementCount = 1;
    const minWidth = 100;
    const maxWidth = 100;
    expect(
      test.calculateWidths(containerWidth, minWidth, maxWidth, elementCount)
    ).toEqual({
      tabWidth: 0,
      slices: 0,
    });
  });
});
