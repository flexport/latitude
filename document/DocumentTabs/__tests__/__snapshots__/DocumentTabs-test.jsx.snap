// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentTabs renders 1`] = `
<div
  className="container_t0nwzw-o_O-hidden_gxjeqd"
  style={
    Object {
      "display": "flex",
    }
  }
>
  <div
    className="tabsContainer_3yo7qj"
  >
    <DocumentTab
      isInDropdown={false}
      isOverflow={false}
      isSelected={true}
      key="0"
      onClick={[Function]}
      onDelete={[Function]}
      size="m"
      width={0}
    >
      doc1.pdf
    </DocumentTab>
    <MoreTab
      selectedLabel="doc1.pdf"
      size="m"
      width={0}
    >
      <DocumentTab
        isInDropdown={true}
        isOverflow={true}
        isSelected={false}
        key="1"
        onClick={[Function]}
        onDelete={[Function]}
        size="m"
        width={0}
      >
        doc2.pdf
      </DocumentTab>
    </MoreTab>
  </div>
  <div
    className="addButtonWrapper_h2lsf5"
  >
    <IconButton
      disabled={false}
      height={
        Object {
          "type": "fixed",
        }
      }
      iconAlignment="left"
      iconName="upload"
      intent="none"
      isLoading={false}
      kind="bare"
      onClick={[Function]}
      size="m"
      type="button"
      width="responsive"
    />
  </div>
</div>
`;
